### ISSUE TITLE
explicit control of the build order

### ISSUE THREAD

**Issue by qhuo** [ninja-build/ninja#232]
It seems to me that all ready to build Edges are contained in a std::set<Edge*> (Plan::ready_). That is, they are sorted by the pointer value comparison. In mostly cases, the order comes out in line with the order build commands are read and parsed from ninja files. And this is ok for us as we generate projects in order of their dependency orders. 

However, sometimes, it may happen that a cpp file from a lower level project is not schedules to compile until much later into the build process, and therefore blocking up all the linking actions, making the total build time unnecessarily long. This typically happens with a clean build, when at the start of ninja, thousands of cpp files are ready to be compiled.

I propose we add some explicit control of the build order for ready targets. Let's say every Edge object should have a member of buildorder, then Plan::ready_ can be changed to set with a compare function, which compare the buildoder member of edges.

There are a few options to determine the build order value for the edges. By default, ninja can assign a sequential number for every Edge object created. We may also introduce a build_order variable to build commands, therefore giveing the generator full control of the order.

A further benefit when above is acomplished, we can indroduce a ninja option that generates randomised build order for all edges. This would be useful to find subtle dependency problems in generated ninja files.

**Comment by Code7R** [comment]
Is there any chance of having any such configuration option implemented in a future version? I see people explaining this wish every couple of years, and sometimes there are some workarounds which improve the behavior.

But there is still no proper solution which would give the developer the explicit control! What I mean is a way of letting the user include some hints to simply tell "prioritize this target and all dependencies in the critical path".

Background: for us the situation looks quite similar to the one in pictures of https://github.com/ninja-build/ninja/issues/376 . I.e. object files from dozens of different (mostly unrelated) targets take ~10min to compile and then the linkers (for just a couple of heavy executables) runs for another 20 minutes. And having a way to prepone those steps by about 8 minutes (i.e. when we could use all CPU cores to build just their object dependencies and not some random ones) would be a great time saver.

And this also does not necessarily have to use historical data or other hacks. The user might already know where the bottleneck is, and could specify the hints manually.

**Comment by mathstuf** [comment]
> This would be useful to find subtle dependency problems in generated ninja files.

The way to fix this is actually to do:

```python
ninja = load_ninja_file()
for output in ninja.outputs:
    ninja.build(target=output) # if this fails, we have missing dependencies somewhere in its deptree
    ninja.clean()
```

With something like `ccache` backing the build, this shouldn't be *too* expensive to run occasionally (say, monthly) to catch issues that creep in. See where I used it to debug `scipy` dependency problems in [this comment](https://github.com/scipy/scipy/issues/18909#issuecomment-2574991551).

> We can even do smarter things like use the build log to estimate the total time of dependent work. (But parallelization and build graphs that aren't simple trees can make these numbers a bit harder...)

Cache hits and misses can also distort such metrics as well.

I'm of the opinion that using build graph depth is probably the easiest first step. Exposing a way for generators to influence it can come later if other ninja-only metrics (graph topology and `ninja_log`) prove inadequate (I can't say that CMake really knows any better how to prioritize any given rule either; probably better to focus on things like [removing](https://gitlab.kitware.com/cmake/cmake/-/issues/15555) [unnecessary](https://gitlab.kitware.com/cmake/cmake/-/issues/25370) [edges](https://gitlab.kitware.com/cmake/cmake/-/issues/17097) in the build graph).

**Comment by eli-schwartz** [comment]
I think ultimately if we want to actually solve this problem to the satisfaction of actual users, we can't beat around the bush about it.

What users want: some project has many targets, only one or a handful of which take 8+ minutes to build and can only use a single core to do it -- and nothing depends on them either, so they will naturally end up as one of the very last edges to *start*. This should somehow "work", and importantly, work in CI which always does scratch builds.

There's no way to *guess* this information. You can't get past the "it's our first time building the project" barrier.

The authors of the build file would need to be able to set a variable in build.ninja for a particular target, setting its priority, e.g. all targets start off with an implicit priority of `buildorder_priority = 50` by default and the lower the number, the sooner the target builds when given a choice between two edges that are ready to build. (Since the sole goal here is to order long-running jobs first and it doesn't "matter" otherwise what order anything is ever built in, I would also think it's better to automatically propagate the priority of a target in the active build graph to any pending dependencies, in order to unlock the high-priority job as soon as possible.)

**Comment by digit-google** [comment]
Yes, that should be rather simple to implement today since currently we assign a weight of 1 to non-phony edges, and 0 to phony ones, then propagate (sum) the counts along the dependency graph. See `EdgeWeightHeuristic()` in `src/build.cc` which is called from  from `Plan::ComputeCriticalPath()`.

I don't have time to work on this, but feel free to experiment modifying this function to read an optional value from the each Edge instance, and see how that works in practice.
